x-pack/plugins/cases/server/services/user_actions/builder.ts

```diff
@@ -0,0 +1,398 @@
+  CaseExternalServiceBasic,
+  noneConnectorId,
+  CaseSettings,
+  CaseConnector,
+  CasePostRequest,
+  CommentUserAction,
+  ConnectorUserAction,
+  ConnectorUserActionWithoutConnectorId,
+  CreateCaseUserActionWithoutConnectorId,
+  DeleteCaseUserAction,
+  DescriptionUserAction,
+  PushedUserAction,
+  PushedUserActionWithoutConnectorId,
+  SettingsUserAction,
+  StatusUserAction,
+  TagsUserAction,
+  TitleUserAction,
+} from '../../../common/api';
+
+import {
+  CASE_REF_NAME,
+  COMMENT_REF_NAME,
+  CONNECTOR_ID_REFERENCE_NAME,
+  PUSH_CONNECTOR_ID_REFERENCE_NAME,
+  SUB_CASE_REF_NAME,
+} from '../../common/constants';
+import { ACTION_SAVED_OBJECT_TYPE } from '../../../../actions/server';
+import { CommonArguments } from './types';
+
+interface BuilderReturnValue<T> {
+  attributes: T;
+  references: SavedObjectReference[];
+}
+
+type CommonBuilderArguments = CommonArguments & {
+  action: UserAction;
+  type: UserActionTypes;
+  value: unknown;
+  valueKey: string;
+  extraReferences?: SavedObjectReference[];
+};
+
+type BuilderFieldTypes = UserActionTypes;
+
+// Title
+type TitleBuilderArgs = { title: string } & CommonArguments;
+type TitleBuilderResponse = BuilderReturnValue<TitleUserAction>;
+
+// Description
+type DescriptionBuilderArgs = { description: string } & CommonArguments;
+type DescriptionBuilderResponse = BuilderReturnValue<DescriptionUserAction>;
+
+// Status
+type StatusBuilderArgs = { status: CaseStatuses } & CommonArguments;
+type StatusBuilderResponse = BuilderReturnValue<StatusUserAction>;
+
+// Tags
+type TagsBuilderArgs = { action: UserAction; tags: string[] } & CommonArguments;
+type TagsBuilderResponse = BuilderReturnValue<TagsUserAction>;
+
+// Pushed
+type PushedBuilderArgs = {
+  externalService: PushedUserAction['payload']['externalService'];
+} & CommonArguments;
+type PushedBuilderResponse = BuilderReturnValue<PushedUserActionWithoutConnectorId>;
+
+// Settings
+type SettingsBuilderArgs = { settings: CaseSettings } & CommonArguments;
+type SettingsBuilderResponse = BuilderReturnValue<SettingsUserAction>;
+
+// Attachments
+type AttachmentBuilderArgs = {
+  action: UserAction;
+  comment: CommentUserAction['payload']['comment'];
+  attachmentId: string;
+} & CommonArguments;
+type AttachmentBuilderResponse = BuilderReturnValue<CommentUserAction>;
+
+// Connector
+type ConnectorBuilderArgs = {
+  connector: ConnectorUserAction['payload']['connector'];
+} & CommonArguments;
+type ConnectorBuilderResponse = BuilderReturnValue<ConnectorUserActionWithoutConnectorId>;
+
+// Create case
+type CreateCaseBuilderArgs = {
+  payload: CasePostRequest;
+} & CommonArguments;
+type CreateCaseBuilderResponse = BuilderReturnValue<CreateCaseUserActionWithoutConnectorId>;
+
+// Create case
+type DeleteCaseBuilderArgs = { connectorId: string } & CommonArguments;
+type DeleteCaseBuilderResponse = BuilderReturnValue<DeleteCaseUserAction>;
+
+type BuilderArgumentsType<T extends BuilderFieldTypes> = T extends 'title'
```

Comment:
Nice work on this class! I see that we need this because of the `getUserActionItemByDifference` function. I wonder if there's a way we can just have an array of fields (title, tags, status, etc) and then build these types dynamically 🤔 , that way we only have to create new functions in this file instead of a new function, and multiple types, and a function in the service itself. I'll do some digging but I did find this: https://stackoverflow.com/questions/63995567/typescript-dynamically-create-types-based-on-object

🤞 we can do something similar

Reply:
If possible we gonna address it on a different PR.